[DG-28] docs: updated Tutorial_v1 (#18)
immutable/imx-docs: docs/tutorials/zero-to-hero-nft-minting.md

---
title: "NFT Minting Tutorial"

> slug: "/zero-to-hero-nft-minting"

Reviewer: Should we update this slug (which will be the URL path) to reflect the name of this tutorial? A downside to this is any links that we can't update (ie. from external websites) will have a broken link.

Another option to consider for later: Should we keep the 'Zero to Hero' theme, and have a section for these sorts of tutorials in our docs site?

Author: Slug updated - good point regarding the theme, will consider as we add more

Reviewer 2: We shouldn't update slugs; if we do, we should set a redirect. See an example of a page-level redirect here: https://github.com/immutable/imx-docs/blob/main/src/pages/docs.tsx - this makes sure https://docs.x.immutable.com/docs redirects to https://docs.x.immutable.com/docs/welcome

Author: Any issue if we update title without updating slug?

Reviewer: Nope, let's not worry about this for now.

sidebar_position: 1
---
*Estimated completion time: 20 mins*

> This tutorial provides a step by step guide on how to mint a NFT on Immutable X. It is designed for developers building on Web3 for the first time, so anyone can follow along regardless of prior experience. This has been simplified for educational purposes and we are working on expanding the content. If you'd prefer to dive into the code, check out our [SDK](https://github.com/immutable/imx-core-sdk/). 

Reviewer: > If you'd prefer to dive into the code, check out our [SDK](https://github.com/immutable/imx-core-sdk/). 

This sentence makes it seem like our SDK will enable them to perform the activity that this tutorial is a guide for, but that's not the case? If the goal is just to direct them to our SDK to do other things, then let's put this line elsewhere and make it clear that using the SDK is not a substitute for this tutorial.

Author: Removed this line for now

Author: Good point - I have removed this line for now.

> This tutorial provides a step by step guide on how to mint a NFT on Immutable X. It is designed for developers building on Web3 for the first time, so anyone can follow along regardless of prior experience. This has been simplified for educational purposes and we are working on expanding the content. If you'd prefer to dive into the code, check out our [SDK](https://github.com/immutable/imx-core-sdk/). 

Reviewer: Something like:
```suggestion
This tutorial provides a step by step guide on how to create and deploy an NFT collection and mint an NFT from that collection on Immutable X. It is designed for developers building on Web3 for the first time, so anyone can follow along regardless of prior experience. This has been simplified for educational purposes and we are working on expanding the content. If you'd prefer to dive into the code, check out our [SDK](https://github.com/immutable/imx-core-sdk/). 
```

Author: Too verbose imo - let's discuss though - want to make sure I am conveying the right message

Reviewer: I guess I just wanted to convey that this tutorial teaches someone to do more than just mint an NFT. Or do you think that is the end goal and everything else is just to facilitate that (and so not relevant)?

But yes, agree that what I proposed could be less verbose.

Author: Agreed - there are a few more things than minting an NFT here. I think for me I want to highlight the end goal in the first sentence, and then in lines 13,14,15 we can outline some of the other activities involved

By the end of this tutorial you will have:
- Prepared an NFT collection

> - Deployed a smart contract

Reviewer: Should we make it clear that the smart contract is for the NFT collection? Just if the user is a total newbie and doesn't see the link between the first and second points.

Author: Agreed - updated

Author: Added

- Minted and listed an NFT 

If you get stuck at any point, reach out on the dev-faq and dev-discussion channels in our [Discord](https://discord.gg/TkVumkJ9D6). Click [here](https://docs.google.com/forms/d/e/1FAIpQLSdTLIXldLRZQB4i2YTHtQwxmrDbTkHphuxtLoVe7j-YVU7VYw/viewform) to provide feedback on the tutorial or let us know what topics you'd like to see in our documentation.

## Step 1: Prerequisites 
There are a few tools required for this tutorial: 

** [Homebrew](https://brew.sh/) **

:::info 
Homebrew is only needed for Mac OS.
:::

> Homebrew installs the packages needed for this tutorial. Copy the command on the Homebrew website, then run in the terminal. 

Reviewer: Let's add a link to the Homebrew website

Author: Added

---
title: "NFT Minting Tutorial"
slug: "/zero-to-hero-nft-minting"
sidebar_position: 1
---
*Estimated completion time: 20 mins*

This tutorial provides a step by step guide on how to mint an NFT on Immutable X. It is designed for developers building on Web3 for the first time, so anyone can follow along regardless of prior experience. This has been simplified for educational purposes and we are working on expanding the content. 

By the end of this tutorial you will have:

- Prepared an NFT collection
- Deployed a smart contract for your NFT collection 
- Minted and listed an NFT 

If you get stuck at any point, reach out on the dev-faq and dev-discussion channels in our [Discord](https://discord.gg/TkVumkJ9D6). Click [here](https://docs.google.com/forms/d/e/1FAIpQLSdTLIXldLRZQB4i2YTHtQwxmrDbTkHphuxtLoVe7j-YVU7VYw/viewform) to provide feedback on the tutorial or let us know what topics you'd like to see in our documentation.

## Step 1: Prerequisites 
There are a few tools required for this tutorial: 

> ** [Homebrew](.https://brewsh/) **

Reviewer 3: The leading `.` is breaking the link - typo?

brew install node@16
``` 

** Yarn**

Yarn is an open source package manager. Follow the steps below to install: 

For **PC** users, run the following command in powershell.
```shell
npm install yarn -g
```

For **Mac** users, run the following command in the terminal.
```shell
brew install yarn
```

**[Visual Studio Code](https://code.visualstudio.com/)
**

Visual Studio Code is the code editor we will be using for this tutorial.

## Step 2: Setup MetaMask 
To trade cryptocurrencies and NFTs, we need a wallet. For this tutorial, we will use Metamask. 

1. Open [metamask.io](https://metamask.io/download/) to install the browser extension.
2. Follow the steps in the plugin to create a new wallet, then record and store your seed phrase in a safe location.

> 3. Ensure you display test networks

Reviewer: Should we provide instructions on how to do this?

Author: I've opted to skip instructions on this one since it is a one click download after pressing the link

4. Change the network selection from **Ethereum Mainnet** to **Ropsten Test Network**.
5. Note down your [private key](https://metamask.zendesk.com/hc/en-us/articles/360015289632-How-to-export-an-account-s-private-key) and [public key ](https://metamask.zendesk.com/hc/en-us/articles/360015289512-How-to-copy-your-MetaMask-account-public-address-)

Changing the network enables us to deploy on a testnet where we can experiment using test Eth. To learn more about transactions on Ethereum work, check out [this page](https://ethereum.org/en/developers/docs/transactions/) from the Ethereum foundation.

## Step 3: Obtain rEth
To deploy our smart contract to the Ropsten test network, we’ll need some test Eth. To get test Eth you can go to the [Ropsten faucet](https://faucet.metamask.io/) and enter your Ropsten account address, then click “Send Ropsten Eth.” It may take a few minutes for the rEth to arrive. 

> If this faucet doesn't work, you can try alternate faucets here:

Reviewer: ```suggestion
If this faucet doesn't work, you can try these alternative faucets:
```

- [Faucet 3 (Request L1Eth)](https://imxfaucet.xyz/)

## Step 4: Setup Pinata
Pinata is a service that allows users to host files on the [InterPlanetary File System](https://docs.ipfs.io/concepts/what-is-ipfs/#decentralization) (IPFS) network. We use Pinata to store our NFT metadata as it ensures the authenticity of the file will be verifiable and the file will always be accessible. Follow the steps below to prepare your collection: 

:::info 
You can use the free version for this tutorial, however consider creating a paid account for your real collection to ensure your images can be loaded. Be aware that your images will be publicly hosted on IPFS. 
:::

1. Sign up to [Pinata](https://www.pinata.cloud/)
2. Prepare 3 images for your NFTs 
3. Upload an image by pressing Upload and selecting the file
   

![Pinata1](../../static/img/zero-to-hero/Pinata1.png 'Pinata1')

4. Upload the remaining 2 images
5. Note down the URL for each file

![Pinata2](../../static/img/zero-to-hero/Pinata2.png 'Pinata2')

Your URLs should have the format `https://gateway.pinata.cloud/ipfs/QmWfjs6CVu4ENgXGNfdPhgaSdzEhCsCfB5XEFUosPFGsNV`

## Step 5: Create Metadata

> [NFT Metadata](https://docs.x.immutable.com/docs/asset-metadata) contains information about the the characteristics and properties of an NFT. 

Reviewer 3: ```suggestion
[NFT Metadata](https://docs.x.immutable.com/docs/asset-metadata) contains information about the characteristics and properties of an NFT. 
```
```suggestion
[NFT Metadata](https://docs.x.immutable.com/docs/asset-metadata) contains information about the characteristics and properties of an NFT. 
```

> [NFT Metadata](https://docs.x.immutable.com/docs/asset-metadata) contains information about the the characteristics and properties of an NFT. 

Reviewer 3: ```suggestion
[NFT Metadata](https://docs.x.immutable.com/docs/asset-metadata) contains information about the characteristics and properties of an NFT. 
```

Click on the folder name to obtain your ** metadata API URL**. This endpoint is used by the metadata crawler to add metadata to your NFTs when they are minted. Note this down. 

![Folder Appearance](../../static/img/zero-to-hero/Step5_Metadata.png 'Folder Appearance')

The resulting URL should have the format `https://gateway.pinata.cloud/ipfs/QmWfKt2pXLnQ2AB5jfS2KYB9K2hxFtgcMNxyndkSGT3yuj`. 

:::info Check your work
Click on your metadata API URL and ensure that there are 3 files named 1,2 & 3. Click into each file and confirm the data is correct. If any of them are empty, you will need to reupload your folder.
:::

## Step 6: Create Etherscan API

An Etherscan API Key is necessary to verify that you're the owner of the smart contract that you're trying to publish. Follow the steps below: 

1. Navigate to [Etherscan](https://etherscan.io/)
2. Sign in (or create a new account) 
3. Navigate to `API-KEYS` and add a new key
4. Note down the generated API key. 

![Etherscan API Key](../../static/img/zero-to-hero/Etherscan1%20no%20arrow.png 'Etherscan API Key')
![Etherscan API Key2](../../static/img/zero-to-hero/Etherscan2.png 'Etherscan API Key')

> ## Step 7: Rename .env File and Update Variables

Reviewer: ```suggestion
## Step 7: Creating our NFT contract
```

Next we will need to install packages needed for deploying our smart contract. 

1. Right click on any file in the explorer
2. Select 'Open in Integrated Terminal'

![Open the integrated terminal](../../static/img/zero-to-hero/Integrated%20Terminal.png 'Open the integrated terminal')

3. Run `npm install --include=dev` and wait until the installation is completed. 
4. Save your work

## Step 9: Deploy Contract
Before we can mint on Layer 2, we need to deploy a smart contract on Layer 1 to ensure assets can be withdrawn to Layer 1. Click [here](https://docs.x.immutable.com/docs/ethereum-scalability) to learn more about the differences between Layer 1 and Layer 2. 

1. Run `yarn hardhat run deploy/asset.ts --network ropsten`
2. It will take ~5 minutes to deploy your contract to Ropsten Etherscan 
3. Copy the deployed contract address 

![Step9_Deploying the contract](../../static/img/zero-to-hero/Step9DeployContract.png 'Deploying the contract')

:::info Check your work
Paste your contract address into [Ropsten Etherscan](https://ropsten.etherscan.io/). It should say **contract** in the upper left. If this says address, ensure you are on the correct network
:::

> ## Step 10: Onboard Smart Contract

Reviewer: ```suggestion
## Step 10: Add your NFT collection to Immutable X
```

The next steps cover how to [onboard](https://docs.x.immutable.com/docs/onboarding) a smart contract with Immutable X to ensure your collection displays correctly:

1. Download the [imx-examples repo](https://github.com/immutable/imx-examples)
2. Open the folder in Visual Studio Code
3. Rename `.env.example` to `.env` and save 
4. Fill in the following: 

<table>
  <thead>
  <tr>
    <th>
      Field Name
    </th>
    <th>
      Description
    </th>
  </tr>
  </thead>
  <tbody>
      <tr>
    <td>
      <code>ALCHEMY_API_KEY</code>
    </td>
    <td>
      <code>AlCHEMY_API_KEY</code> from step 7
    </td>
  </tr>
  <tr>
    <td>
      <code>OWNER_ACCOUNT_PRIVATE_KEY</code>
    </td>
    <td>
     Metamask private key  from step 2
    </td>
  </tr>
  <tr>
    <td>
      <code>COLLECTION_CONTRACT_ADDRESS</code>
    </td>
    <td>
       Deployed contract address from previous step
    </td>
  
  </tr>
  </tbody>
</table>

![Step10_1](../../static/img/zero-to-hero/Step10_1.png 'Step10_1')

1. Save 
2. Run `npm install`

> ## Step 11: Register User

Reviewer: ```suggestion
## Step 11: Register as a user on Immutable X
```

Author: Done

> Registering as a user creates an Layer 2 wallet and enables us to make transactions on Layer 2. 

Reviewer: Can we add something about how it is imperative that they become an Immutable X user in order to add a collection?

Author: Done

To add metadata schema to the collection, run the command `npm run onboarding:add-metadata-schema` in the integrated terminal. 

## Step 15: Mint NFT

> Now that we have onboarded our contract, the final step is to add our assets to the blockchain by minting them. 

Reviewer: ```suggestion
Now that we have added contract to Immutable X, the final step is to add our assets to the blockchain by minting them. 
```

Author: Done
